spec/compiler/normalize/def_spec.cr

```diff
@@ -231,5 +231,25 @@ module Crystal
       other_def = a_def.expand_default_arguments(Program.new, 0, ["abstract"])
       other_def.to_s.should eq("def foo:abstract(abstract __arg0)\n  options = {}\n  @abstract = __arg0\nend")
     end
+
+    describe "gives correct body location with" do
+      {"default arg"                  => "def testing(foo = 5)",
+       "default arg with restriction" => "def testing(foo : Int = 5)",
+       "splat arg"                    => "def testing(*foo : Array)",
+       "block instance var arg"       => "def testing(&@foo : ->)",
+      }.each do |(suf, definition)|
+        describe suf + "," do
+          {"with body" => "zzz = 7\n", "without body" => ""}.each do |(suf, body)|
+            it suf do
+              a_def = parse("#{definition}\n#{body}end").as(Def)
+              actual = a_def.expand_default_arguments(Program.new, 1)
+
+              actual.location.should eq Location.new("", line_number: 1, column_number: 1)
+              actual.body.location.should eq Location.new("", line_number: 2, column_number: 1)
+            end
+          end
+        end
+      end
+    end
```

Comment:
As a general consideration, I don't like specs in iterations because they seem unnecessarily hard to understand. Especially with nested iterations.
My preferred style would be a spec helper called with explicit source code of each combination. That also offers much flexibility and it would be easy to add specs for multi-line parameter lists for example.
I'm happy to accept this as is though.

Reply:
I don't mind, but in this particular case I don't see a good way to write this with a helper, particularly due to the nested iterations.
I think I will not do this change.

Reply:
This would be my proposal: https://github.com/straight-shoota/crystal/commit/5da588658743f719c9318c7e5c52e35854ff7494

Reply:
Hmm some things there that make me like that one less:
* The `with body` context comes in the wrong order (precedence).
* No possibility of auto-indenting the different cases.
* It is indeed a lot more code, and it's not like the helper will be reused for different tests.

For now I tweaked my version, maybe it looks a little better

Reply:
[Diff between the two is still available](https://github.com/straight-shoota/crystal/compare/645a241d38bf6e3850b7bf867d549dad7f382346..5da588658743f719c9318c7e5c52e35854ff7494)